include/volume/volume_cooling_nonspherical_gaussians_crhmc.hpp

```diff
@@ -0,0 +1,612 @@
+
+    Func f(f_params);
+    Grad g(f_params);
+    Hess h(f_params);
+    ZeroFunctor<Point> zerof;
+
+    Input input = convert2crhmc_input<Input, Polytope, Func, Grad, ZeroFunctor<Point>>(P, f, g, zerof);
+
+    //Input input = convert2crhmc_input<Input, Polytope, Func, Grad, Hess>(P, f, g, h);
+
+    typedef crhmc_problem<Point, Input> CrhmcProblem;
+    
+    CrhmcProblem problem = CrhmcProblem(input);
+
+    if(problem.terminate) { return 0;}
+
+    problem.options.simdLen = simdLen;
+
+    crhmc_walk_params params(input.df, p.dimension(), problem.options);
+
+    // Create the walk object for this problem
+    CRHMCWalkType walk = CRHMCWalkType(problem, p, input.df, input.f, params);
+
+    burnIn<CRHMCWalkType, crhmc_walk_params, simdLen, Grad, Func, CrhmcProblem>(
+        p, walk_length, rng, g, f, params, problem, walk, 2.5 * N);
+
+    NT last_a = a;
+    NT last_ratio = 0.1;
+    NT k = 1.0;
+    const NT tol = 0.00001;
+    bool done = false;
+    std::vector<NT> fn(N, NT(0.0));
+    std::list<Point> randPoints;
+
+    // sample N points
+    PushBackWalkPolicy push_back_policy;
+    bool raw_output = false;
+    typedef CrhmcRandomPointGenerator<CRHMCWalkType> CRHMCRandomPointGenerator;
+
+    CRHMCRandomPointGenerator::apply(problem, p, N, walk_length, randPoints,
+                                    push_back_policy, rng, g, f, params, walk, simdLen, raw_output);
+
+    while (!done) {
+        NT new_a = last_a * std::pow(ratio, k);
+
+        auto fnit = fn.begin();
+        for (auto pit = randPoints.begin(); pit != randPoints.end(); ++pit, fnit++) {
+            *fnit = eval_exp(*pit, inv_covariance_matrix, new_a, last_a);
+        }
+
+        std::pair<NT, NT> mv = get_mean_variance(fn);
+
+        // Compute a_{i+1}
+        if (mv.second / (mv.first * mv.first) >= C || mv.first / last_ratio < 1.0 + tol) {
+            if (k != 1.0) {
+                k = k / 2;
+            }
+            done = true;
+        } else {
+            k = 2 * k;
+        }
+
+        last_ratio = mv.first;
+    }
+
+    // Return the new a value as a scalar
+    return last_a * std::pow(ratio, k);
+}
+
+// Compute the sequence of non spherical gaussians
+template<
+    int simdLen,
+    typename Point,
+    typename Polytope,
+    typename NT,
+    typename MT,
+    typename RandomNumberGenerator
+>
+void compute_annealing_schedule(Polytope Pin_copy,
+                                Polytope P_copy,
+                                NT const& ratio,
+                                NT const& C,
+                                NT const& frac,
+                                unsigned int const& N,
+                                unsigned int const& walk_length,
+                                NT const& error,
+                                std::vector<NT>& a_vals,
+                                MT const& inv_covariance_matrix,
+                                RandomNumberGenerator& rng) 
```

Comment:
Is `rng` an input for the spherical gaussians as well?

Reply:
Yes: https://github.com/GeomScale/volesti/blob/d424f3e31817cf33099a903ca42706c16627cb4d/include/volume/volume_cooling_gaussians.hpp#L198